### ISSUE TITLE
VectorStore interface should be split, with readonly operations becoming a super interface

### ISSUE THREAD

**Issue by johnsonr** [spring-projects/spring-ai#1290]
The `VectorStore` interfaces covers two things: similarity search and `DocumentWriter`/add/accept/delete.

These are different concepts and a `SimilaritySearchService` or similarly named interface should be split out.

I currently have something I would like to expose as a `VectorStore` that is readonly and does not support the ingestion operations.

**Comment by markpollack** [comment]
Yes, I agree as well that we need some redesign here and also te ability to pass options to add/delete operations.  There are also other operations besides similarity search that need to be exposed.

**Comment by rjrudin** [comment]
@markpollack Let me know if this worth a separate ticket, happy to open one - I'd also like to provide additional context to a `SearchRequest`, such as additional query context that isn't part of the user's query. For example, a user may wish to engage a chatbot about crimes in a particular area defined by a bounding box. Under the hood, I want to use that bounding box to select a subset of records in a database and then do a similarity search on those records. That additional query context doesn't fit into a `Filter.Expression` - perhaps a metadata map on `SearchRequest` would be a reasonable extension point?

**Comment by ThomasVitale** [comment]
As part of the new (experimental) Modular RAG features, a VectorStoreDocumentRetriever has been introduced to support search operations decoupled from the generic CRUD operations in the VectorStore API.

@johnsonr does that help with your use case?

Docs: https://docs.spring.io/spring-ai/reference/api/retrieval-augmented-generation.html#_retrieval

**Comment by markpollack** [comment]
@sobychacko thoughts on this.

**Comment by markpollack** [comment]
Yep.  We can retrofit a `SimilaritySearchOperations` interface (name TBD) to that it can be passed around to code without fear of that code adding anything.

**Comment by markpollack** [comment]
@rjrudin The `SearchRequest` is specifically to feed in the three  things that are required for a similarity search - the query string, topk to return and filter expression as a passthrough to execution via a vector db that supports similarity search. Your query that you mention seems orthogonal to that execution path/use case.  It sounds more like a geo-query (great with mongodb) in a non-vector-store database and then some sort of custom ranking of those records (perhaps by closest to a point).
